src/java.base/share/classes/jdk/internal/foreign/MemorySegmentRenderUtil.java

```diff
@@ -0,0 +1,367 @@
+    }
+
+    public static void renderView(MemorySegment memorySegment,
+                                  MemoryLayout memoryLayout,
+                                  Consumer<? super CharSequence> action,
+                                  ViewState state,
+                                  String suffix) {
+
+        // TODO: Replace with "patterns in switch statement" once this becomes available.
+
+        if (memoryLayout instanceof ValueLayout.OfByte ofByte) {
+            action.accept(renderValueLayout(state, ofByte, Byte.toString(memorySegment.get(ofByte, state.indexAndAdd(ofByte))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfShort ofShort) {
+            action.accept(renderValueLayout(state, ofShort, Short.toString(memorySegment.get(ofShort, state.indexAndAdd(ofShort))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfInt ofInt) {
+            action.accept(renderValueLayout(state, ofInt, Integer.toString(memorySegment.get(ofInt, state.indexAndAdd(ofInt))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfLong ofLong) {
+            action.accept(renderValueLayout(state, ofLong, Long.toString(memorySegment.get(ofLong, state.indexAndAdd(ofLong))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfFloat ofFloat) {
+            action.accept(renderValueLayout(state, ofFloat, Float.toString(memorySegment.get(ofFloat, state.indexAndAdd(ofFloat))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfDouble ofDouble) {
+            action.accept(renderValueLayout(state, ofDouble, Double.toString(memorySegment.get(ofDouble, state.indexAndAdd(ofDouble))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfChar ofChar) {
+            action.accept(renderValueLayout(state, ofChar, Character.toString(memorySegment.get(ofChar, state.indexAndAdd(ofChar))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfBoolean ofBoolean) {
+            action.accept(renderValueLayout(state, ofBoolean, Boolean.toString(memorySegment.get(ofBoolean, state.indexAndAdd(ofBoolean))), suffix));
+            return;
+        }
+        if (memoryLayout instanceof ValueLayout.OfAddress ofAddress) {
+            action.accept(renderValueLayout(state, ofAddress, memorySegment.get(ofAddress, state.indexAndAdd(ofAddress)).toString(), suffix));
+            return;
+        }
+        if (memoryLayout instanceof PaddingLayout paddingLayout) {
+            action.accept(state.indentSpaces() + paddingLayout.bitSize() + " padding bits");
+            state.indexAndAdd(paddingLayout);
+            return;
+        }
+        if (memoryLayout instanceof GroupLayout groupLayout) {
+
+            // Strictly, we should provide all permutations of nested unions.
```

Comment:
Not sure I get this comment

Reply:
So, suppose there is a nested union such that: U = (A|B)|(C|D). Then the actual layout U could be either of AC, AD, BC or BD. Hence, we could present all these combinations.

Reply:
Ah I see. I think that, perhaps, it would be better to collect all fields in a union, and present them all (like if it was a struct). I don't see any particular problem with that, and I don't think the API should "pick" a field.

Reply:
Ah ok, I misread your comment. I think what the impl does now is actually sensible, as it presents the full `(A|B)|(C|D)`.